core/domain/story_domain.py

```diff
@@ -317,82 +337,72 @@ def validate(self):
-        # The following 2 loops create the prerequisite skill id list which
-        # contains all the prerequisite skill ids of the story nodes minus
-        # its intersection with the acquired skill id list.
-        for node in self.nodes:
-            for skill_id in node.prerequisite_skill_ids:
-                prerequisite_skill_ids.append(skill_id)
-        for node in self.nodes:
-            for skill_id in node.acquired_skill_ids:
-                if skill_id in prerequisite_skill_ids:
-                    prerequisite_skill_ids.remove(skill_id)
-
-        # The following loop calculates the nodes that are unlocked initially
-        # based on the prerequisite skill ids calculated above.
-        for node in self.nodes:
-            node_unlocked = True
-            for skill_id in node.prerequisite_skill_ids:
-                if skill_id not in prerequisite_skill_ids:
-                    node_unlocked = False
-                    break
-            if node_unlocked:
-                unlocked_nodes.append(node)
-        print unlocked_nodes
-        # The while loop is there so that, is a node is unlocked, but its
-        # prerequisite skills are acquired at a later stage, then that node
-        # needs to be checked again. (The second test in
-        # test_all_nodes_visited()) checks this.
-        while len(visited_nodes) != previous_visited_nodes_length:
-            previous_visited_nodes_length = len(visited_nodes)
-            # The following loop iterates through these unlocked nodes and marks
-            # a node as visited if all its prerequisite skills are present in
-            # the list calculated above. Also, new nodes are unlocked are based
-            # on destination node ids of these visited nodes.
-            # All of the initial unlocked nodes would be visited, but for
-            # consistency, that check is also done in this loop.
-            for node in unlocked_nodes:
-                node_visited = True
-                for skill_id in node.prerequisite_skill_ids:
-                    if skill_id not in prerequisite_skill_ids:
-                        node_visited = False
-                        break
-                if node_visited:
-                    if node not in visited_nodes:
-                        visited_nodes.append(node)
-                        for skill_id in node.acquired_skill_ids:
-                            prerequisite_skill_ids.append(skill_id)
-                        for node_id in node.destination_node_ids:
-                            node = self.get_node_by_id(node_id)
-                            unlocked_nodes.append(node)
-
-            for node in visited_nodes:
-                if node in unlocked_nodes:
-                    unlocked_nodes.remove(node)
-
-        if len(visited_nodes) < len(self.nodes):
+        if False in is_node_visited:
             raise utils.ValidationError('Expected all nodes to be reachable.')
```

Comment:
Again perhaps saying which nodes are not.

Reply:
Done.